### PR TITLE
feat: introduce apiserver for operator

### PR DESCRIPTION
**Pull request by zyy17**
## What's changed

Introduce a apiserver for greptimedb-operator. The trust 3rd application can use http service to access the metadata of `GreptimeDBCluster` without using K8s API.

For example:

- `GET /api/v1alpha1/namespaces/all/clusters`: Get all clusters;
- `GET /api/v1alpha1/namespaces/foo/clusters/bar`: Get `bar` cluster of `foo` namespace;

## Summary by CodeRabbit

- **New Features**
	- Introduced an API server for managing GreptimeDBCluster resources with endpoints for cluster retrieval.
	- Added command-line flags to enable the API server and specify its port.

- **Bug Fixes**
	- Improved error handling for API server startup failures.

- **Tests**
	- Added unit tests for the new API server functionality, validating HTTP responses and interactions with a mock Kubernetes client.

### CODE REVIEW

File: pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+
+	// Get a specific cluster in a specific namespace.
+	api.GET("/namespaces/:namespace/clusters/:name", s.getCluster)
+
+	klog.Infof("HTTP service is running on port %d", s.port)
+
+	return router.Run(fmt.Sprintf(":%d", s.port))
+}
+
+func (s *Server) getClusters(c *gin.Context) {
```
Comment (daviderli614) [inline pkg/apiserver/apiserver.go:151]: it be better to use listCluster here ?

File: pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+	} {
+		if err := s.getPods(ctx, namespace, name, kind, topology); err != nil {
+			return nil, err
+		}
+	}
+
+	return topology, nil
+}
+
+func (s *Server) getPods(ctx context.Context, namespace, name string, kind greptimev1alpha1.ComponentKind, topology *GreptimeDBClusterTopology) error {
```
Comment (daviderli614) [inline pkg/apiserver/apiserver.go:290]: It be better to use listPod here.

File: pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+	return &Pod{
+		Name:      pod.Name,
+		Namespace: pod.Namespace,
+		IP:        pod.Status.PodIP,
+		Status:    string(pod.Status.Phase),
+		Node:      pod.Spec.NodeName,
+		Resource:  pod.Spec.Containers[constant.MainContainerIndex].Resources,
+		StartTime: pod.Status.StartTime,
+	}
+}
```
[the remark this answers is not in the document]
Comment (daviderli614) [reply pkg/apiserver/apiserver.go:329]: agree suggestion

File: cmd/operator/app/command.go [hunk 4]
```diff
@@ -99,6 +100,19 @@ func NewOperatorCommand() *cobra.Command {
...
 			}
 
+			if o.EnableAPIServer {
+				server := apiserver.NewServer(mgr.GetClient(), &apiserver.Options{
+					Port: o.APIServerPort,
+				})
+
+				go func() {
+					if err := server.Run(); err != nil {
+						setupLog.Error(err, "unable to run HTTP service")
```
Comment (daviderli614) [inline cmd/operator/app/command.go:110]:
```suggestion
						setupLog.Error(err, "unable to run api service")
```

File: cmd/operator/app/options/options.go [hunk 5]
```diff
@@ -27,17 +27,23 @@ type Options struct {
...
+	EnableAPIServer      bool
+	APIServerPort        int
 }
 
 func NewDefaultOptions() *Options {
 	return &Options{
 		MetricsAddr:     defaultMetricsAddr,
 		HealthProbeAddr: defaultHealthProbeAddr,
+		EnableAPIServer: false,
+		APIServerPort:   8081,
```
Comment (daviderli614) [inline cmd/operator/app/options/options.go:39]: suggestion: Using constant `defaultAPIServerPort` replace 8081.

File: pkg/apiserver/apiserver.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+// Server provides an HTTP API to interact with Greptime CRD resources.
+type Server struct {
+	client.Client
+
+	port int
+}
+
+// Options represents the options for the Server.
+type Options struct {
+	// Port is the port that the HTTP service will listen on.
```
Comment (daviderli614) [inline pkg/apiserver/apiserver.go:48]:
```suggestion
type Options struct {
	// Port is the port that the API service will listen on.
```

File: pkg/apiserver/apiserver.go (hunk 2 above)
Comment (zyy17) [reply pkg/apiserver/apiserver.go:290]: list pods is part of the `getPods()`.

File: pkg/apiserver/apiserver.go (hunk 1 above)
Comment (zyy17) [reply pkg/apiserver/apiserver.go:151]: I don't think so. It's coventional to use `get` in REST API.

File: pkg/apiserver/apiserver.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+		data = append(data, GreptimeDBCluster{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			Spec:      &cluster.Spec,
+			Status:    &cluster.Status,
+			Topology:  topology,
+		})
+	}
+
```
[the remark this answers is not in the document]
Comment (zyy17) [reply pkg/apiserver/apiserver.go:204]: It's not a pointer type in `clusters.Items`.

**Review by daviderli614 — APPROVED** [review]
LGTM

**Pull request merged**